server/graphql/schema.js

```diff
@@ -93,23 +113,77 @@ const resolvers = {
+      });
       await newUser.save();
-      const token = signToken(newUser); 
-      // const token = jwt.sign({ 
-      //   requester: email, 
-      //   iat: Math.floor(Date.now() / 1000)
-      // });// jwt token here// jwt token here
+      const token = signToken(newUser);
       return { token, user: newUser };
     },
-    // update an existing user
-    UPDATE_USER: async (_, { username, email, password }, context) => {
-      const updatedUser = await User.findOneAndUpdate(
-        { username },
-        { email, password },
-        { new: true }
-      );
-      return updatedUser;
+    CHANGE_EMAIL: async (_, { userId, newEmail }, context) => {
+      // authenticate the user
+      if (!context.user || context.user._id !== userId) {
+        throw new Error('Unauthorized');
+      }
+      // check if the new email is already in use
+      const emailExists = await User.findOne({ email: newEmail });
+      if (emailExists) {
+        return { success: false, message: 'Email already in use' };
+      }
+      // find the user and update the email
+      const user = await User.findById(userId);
+      if (!user) {
+        throw new Error('User not found');
+      }
+      user.email = newEmail;
+      await user.save();
+      return { success: true, message: 'Email successfully changed' };
+    },
+    CHANGE_USERNAME: async (_, { userId, newUsername }, context) => {
+      // authenticate the user (ensure they are who they claim to be)
+      if (!context.user || context.user._id !== userId) {
+        throw new Error('Unauthorized');
+      }
+      // check if the new username is already in use
+      const usernameExists = await User.findOne({ username: newUsername });
+      if (usernameExists) {
+        return { success: false, message: 'Username already in use' };
+      }
+      // find the user and update the username
+      const user = await User.findById(userId);
+      if (!user) {
+        throw new Error('User not found');
+      }
+      user.username = newUsername;
+      await user.save();
+      return { success: true, message: 'Username successfully changed' };
+    },
+    CHANGE_PASSWORD: async (_, { userId, oldPassword, newPassword }, context) => {
+      // authenticate the user 
+      if (!context.user || context.user._id !== userId) {
+        return { success: false, message: 'Unauthorized' };
+      }
+      // find the user by ID
+      const user = await User.findById(userId);
+      if (!user) {
+        return { success: false, message: 'User not found' };
+      }
+      // check if the old password is correct
+      const isMatch = await bcrypt.compare(oldPassword, user.password);
+      if (!isMatch) {
+        return { success: false, message: 'Old password is incorrect' };
+      }
+      // hash the new password
+      const hashedPassword = await bcrypt.hash(newPassword, saltRounds);
+      // update the user's password
+      user.password = hashedPassword;
+      await user.save();
```

Comment:
revoke the user token after a successful password change, so that they are forced to log back in (same methodology that Google uses for Workspaces)